doc/compiler/optimizer/InliningAndInliningGuards.md

```diff
@@ -0,0 +1,406 @@
+and more code has been added in Inliner to be incrementally aware of the opportunities when going into the callees.
+
+# 4. Overview Of Guards
+
+## Hot Code Replacement (HCR) Guard
+
+HCR is on by default. Any Java method can be redefined by the user while the program is running. As a result,
+we must be able to reflect the output when any of the redefinitions happen. Even if there is a direct call such
+as `invoke_static` or `invoke_special`, a guard must be added around it to preserve the functional correctness
+in case that the method is redefined. The HCR guard works as a nopable guard (discussed in the vnext section).
+The checking is not done explicitly in the generated code, which is good since there will be no overhead in doing
+the extra check. If HCR is not enabled, a direct call is inlined without any guard.
+
+HCR guards can be treated as a second guard. If there is no other guard, it introduces a new guard. If there is
+a guard, it adds another layer as a second guard. For example, it is like two conditions, a virtual guard and
+a HCR guard both exist in place to indicate if the method is overridden, or if the method is redefined.
+
+## Nopable Guard
+
+Nopable guard is like a conditional test conceptually except there is no explicit testing in the generated code.
+There is an `“if”` in IL, but it will always fall into the inlined code during the compilation. If the user
+redefines the method that is inlined in the future, we will go patch the code where it is inlined and take the
+slow path to invoke the new version of the method. Such events are stop-the-world events where application
+threads are guaranteed to not be executing, and hence patching in a full instruction into the instruction
+stream is safe.
+
+For a nopable guard to be used, it depends on the state of the class hierarchy at compile time. For example, `B`
+subclasses an interface `A` when the compilation is done. `A` declares a method `foo()` which is implemented by
+`B`. When the method is compiled and inlined, there is only one target `B`. In the future, if a subclass `C` is
+loaded after the compilation is done and the code has been generated, there could be an object type of `B` as
+well as type `C` that goes through `A.foo()`. Once `C` gets loaded, nopable guards cannot be used any more.
+Before `C` gets loaded, JIT has an infrastructure that receives a notification that a new class is being loaded.
+It assesses if any of the guards will be invalidated. We will patch the code where we have the nopable guards to
+go to the slow path to invoke the call. Afterwards, the class `C` will be allowed to be loaded and to enter the
+system. When we do nopable guard patching, the guard changes the state. We convert an unconditional fall through
+into an unconditional goto.  Even if it is the object `B` that flows through, it will still jump to the slow
+path to invoke the virtual call. Nopable guards require JIT to monitor the changes of program behaviour such as
```

Comment:
"JIT" -> "JIT compiler"...either way you should find a consistent way to refer to it in the article.

"require JIT" -> "require JIT hook routines" 

"changes of" -> "changes in"